maven-core/src/main/java/org/apache/maven/utils/introspection/ReflectionValueExtractor.java

```diff
@@ -0,0 +1,325 @@
+                            propertyPosition,
+                            tokenizer.getPosition(),
+                            value,
+                            tokenizer.nextToken(INDEXED_END));
+                    break;
+                case MAPPED_START:
+                    value = getMappedValue(
+                            expression,
+                            propertyPosition,
+                            tokenizer.getPosition(),
+                            value,
+                            tokenizer.nextToken(MAPPED_END));
+                    break;
+                case PROPERTY_START:
+                    propertyPosition = tokenizer.getPosition();
+                    value = getPropertyValue(value, tokenizer.nextPropertyName());
+                    break;
+                default:
+                    // could not parse expression
+                    return null;
+            }
+        }
+
+        return value;
+    }
+
+    private static Object getMappedValue(
+            final String expression, final int from, final int to, final Object value, final String key) {
+        if (value == null || key == null) {
+            return null;
+        }
+        if (value instanceof Map) {
+            try {
+                Object[] localParams = new Object[] {key};
+                ClassMap classMap = getClassMap(value.getClass());
+                Method method = classMap.findMethod("get", localParams);
+                return method.invoke(value, localParams);
+            } catch (Exception e) {
+                throw new RuntimeException(e);
+            }
+        } else {
+            final String message = String.format(
+                    "The token '%s' at position '%d' refers to a java.util.Map, but the value seems is an instance of '%s'",
+                    expression.subSequence(from, to), from, value.getClass());
+            throw new IllegalArgumentException(message);
+        }
+    }
+
+    private static Object getIndexedValue(
+            final String expression, final int from, final int to, final Object value, final String indexStr) {
+        int index;
+        try {
+            index = Integer.parseInt(indexStr);
+
+            if (value.getClass().isArray()) {
+                return Array.get(value, index);
+            }
+
+        } catch (NumberFormatException e) {
+            final String message = String.format("The indexStr '%s' is not a number", indexStr);
+            throw new IllegalArgumentException(message);
+        }
+
+        if (value instanceof List) {
+            try {
+                ClassMap classMap = getClassMap(value.getClass());
+                // use get method on List interface
+                Object[] localParams = new Object[] {index};
+                Method method = classMap.findMethod("get", localParams);
+                return method.invoke(value, localParams);
+            } catch (Exception e) {
+                throw new RuntimeException(e);
```

Comment:
we likely shouldn't be throwing a raw runtime exception here. At least use a subclass, but possibly the logic here needs a rethink

Reply:
The original code was returning null.  I've added tests and reverted this code change.